### PR TITLE
Allow caching more than the last operator in the CircuitSampler

### PR DESCRIPTION
**Pull request by Cryoris**
### Summary

Allow caching more than the last operator in the CircuitSampler. This is especially useful in optimization settings where the sampler is used to evaluate the gradient and the loss function in an alternating manner.

### Details and comments

Currently, the `CircuitSampler` caches the last operator it converted to not have to do the following steps again: reduce, extract circuits and transpile circuits. In several applications however we evaluate the same few operators over and over again with different parameters. This is not cached, but could. 

In this PR we add a `caching` argument that is `'last'` by default (cache only the last op, that's the current behaviour) but can be set to `'all'` to cache all operators. 

To reliably and efficiently recognize operators that have been cached, we had to add a unique identifier to the `OperatorBase` in form of a global instance counter (like the quantum circuits and registers) that is used to calculate the hash.

Example 1:
```python
import numpy as np
from qiskit import QuantumCircuit, Aer
from qiskit.circuit import Parameter
from qiskit.opflow import CircuitSampler, Gradient, Y, StateFn

x, y = Parameter('x'), Parameter('y')
circuit = QuantumCircuit(2)
circuit.rx(x, 0)
circuit.ry(y, 1)

loss = ~StateFn(Y) @ StateFn(circuit)
grad = Gradient().convert(loss, [x, y])

bound_loss = loss.bind_parameters({x: 1, y: 1})
bound_grad = grad.bind_parameters({x: 1, y: 1})

backend = Aer.get_backend('statevector_simulator')
sampler = CircuitSampler(backend, caching='all')  # previously only option 'last' was possible

reps = 1000

# alternating evaluation
for _ in range(reps):
    _ = sampler.convert(bound_loss).eval()
    _ = sampler.convert(bound_grad).eval()

# grouped evaluation
for _ in range(reps):
    _ = sampler.convert(bound_loss).eval())

for _ in range(reps):
    _ = sampler.convert(bound_grad).eval()
```
Without this PR (or with `caching='last'`) the runtime of the alternating evaluation took twice as long as the grouped evaluation.

Example 2:
Apply gradient descent to this simple 3 qubit circuit speeds up the optimization by a factor of 3. For larger circuits the speedup becomes even more prominent.
```python
params = ParameterVector('x', 4)
circuit = QuantumCircuit(3)
circuit.ry(np.pi / 4, 0)
circuit.ry(np.pi / 3, 1)
circuit.ry(np.pi / 7, 2)

# V0(theta0, theta1): Parametrized layer 0
circuit.rz(params[0], 0)
circuit.rz(params[1], 1)

# W1: non-parametrized gates
circuit.cx(0, 1)
circuit.cx(1, 2)

# V_1(theta2, theta3): Parametrized layer 1
circuit.ry(params[2], 1)
circuit.rx(params[3], 2)

# W2: non-parametrized gates
circuit.cx(0, 1)
circuit.cx(1, 2)
ansatz = StateFn(circuit)

observable = I ^ I ^ Y

initial_point = np.array([0.432, -0.123, 0.543, 0.233])
```

### CODE REVIEW

File: qiskit/opflow/operator_base.py [hunk 1]
```diff
@@ -136,6 +142,9 @@ def _indent(lines: str, indentation: str = INDENTATION) -> str:
             indented_str = indented_str[:-len(indentation)]
         return indented_str
 
+    def __hash__(self):
```
Comment (woodsp-ibm) [inline qiskit/opflow/operator_base.py:145]: As operator also supports `__eq__` and there is an expected behavior between them (if equal should have same hash) this will break things if they are used in Python structures/containers relying on this. I see in the code above id() was used before - since this is unique for objects with overlapping lifetimes maybe this could/should be used. Though I see the original operator is not stored just a reduced version of it hence the original operator lifetime is not known and not in any way influenced by the cache. In which case maybe change this from`__hash__` to `identifer`, `uid` or some name other than hash.

File: qiskit/opflow/converters/circuit_sampler.py [hunk 2]
```diff
@@ -198,7 +214,17 @@ def replace_circuits_with_dicts(operator, param_index=0):
...
-        if params:
+        # store the operator we constructed, if it isn't stored already
+        if op_id not in self._cached_ops.keys():
+            op_cache = OperatorCache()
+            op_cache.reduced_op_cache = self._reduced_op_cache
+            op_cache.circuit_ops_cache = self._circuit_ops_cache
+            op_cache.transpiled_circ_cache = self._transpiled_circ_cache
+            op_cache.transpile_before_bind = self._transpile_before_bind
+            op_cache.transpiled_circ_templates = self._transpiled_circ_templates
+            self._cached_ops[op_id] = op_cache
```
Comment (woodsp-ibm) [inline qiskit/opflow/converters/circuit_sampler.py:225]: Would we want a method to clear the cache at all? Otherwise I guess, if you are using the same CircuitSampler instance on different occasions, with different operators, this just continues to build up.

File: qiskit/opflow/converters/circuit_sampler.py [hunk 3]
```diff
@@ -147,34 +152,45 @@ def convert(self,
...
-            # Clear caches
-            self._last_op = operator
-            self._reduced_op_cache = None
-            self._circuit_ops_cache = None
-            self._transpiled_circ_cache = None
-            self._transpile_before_bind = True
-
-        if not self._reduced_op_cache:
+        # check if the operator should be cached
+        op_id = hash(operator)
```
Comment (woodsp-ibm) [inline qiskit/opflow/converters/circuit_sampler.py:156]: See the comment around hash in operator base. When I first saw this and thought hash my first thought was that hash collisions can exist which makes this suspect - I saw later its not what I thought it was...

File: qiskit/opflow/operator_base.py (hunk 1 above)
Comment (Cryoris) [reply qiskit/opflow/operator_base.py:145]: Good point, but `id` cannot be used (I tried) because Python may re-use the address of an old "dead" object for a new object -- and then the circuit sampler uses the "dead" cache instead of caching the new one.

File: qiskit/opflow/converters/circuit_sampler.py (hunk 2 above)
Comment (Cryoris) [reply qiskit/opflow/converters/circuit_sampler.py:225]: We can add one, that's a good idea 👍🏻

File: qiskit/opflow/operator_base.py (hunk 1 above)
Comment (woodsp-ibm) [reply qiskit/opflow/operator_base.py:145]: Yes, I understand using id(), as there, may be problematic since the operator itself is not added to the cache, only a reduced version, hence the original my be disposed of and a new object, with that id() replace it. Has the operator been added to the cache, as well as the reduced version, then I think it could have been used. Otherwise, its the hash method you have there, but as a different name I think, i.e. not `__hash__`

Comment (Cryoris) [reply qiskit/opflow/operator_base.py:145]: Ok, we can call it `instance_id` or so 👍🏻

File: qiskit/opflow/converters/circuit_sampler.py (hunk 3 above)
Comment (Cryoris) [reply qiskit/opflow/converters/circuit_sampler.py]: Resolved in b585a0f

**Review by woodsp-ibm — APPROVED** [review]
